File: solutions/fortran/allergies/2/allergies.f90

module allergies
  implicit none
character(len=12), dimension(8), parameter :: allergens = & 
[character(len=12) :: 'eggs', 'peanuts', 'shellfish', &
'strawberries', 'tomatoes', 'chocolate', 'pollen', 'cats']

contains

  logical function allergicTo(allergy_str, allergy_key) 
    character(len=*), intent(in) :: allergy_str
    integer, intent(in) :: allergy_key

    allergicTo = btest(allergy_key, findloc(allergens, allergy_str, 1) - 1 )
   end function



  function allergicList(allergy_key)
    integer, intent(in) :: allergy_key
    character(len=100) :: allergicList
    integer :: i
    allergicList = ' '

    do i = 0,7
      if (btest(allergy_key, i)) then
                allergicList =  trim(allergicList) //" "// trim(allergens(i+1))
        end if
    end do
    allergicList = adjustl(allergicList)
  end function



end module

File: solutions/free-pascal/allergies/1/Allergies.pas
unit Allergies;

{$mode ObjFPC}{$H+}

interface

type
  TStrArray = Array Of String;

function list(const score : integer) : TStrArray;
function AllergicTo(const item : string; const score : integer) : boolean;

implementation

uses SysUtils;

const 
  allAllergens : array[0..7] of string = ('eggs', 'peanuts', 'shellfish', 'strawberries', 'tomatoes', 'chocolate', 'pollen', 'cats');

function list(const score : integer) : TStrArray;
var i,s: integer;
begin
  s := 1;
  result := [];
  for i:= low(allAllergens) to high(allAllergens) do
  begin
    if ((s and score)<>0) then
      Insert(allAllergens[i], result, length(result));
    s := s shl 1;
  end;
end;

function AllergicTo(const item : string; const score : integer) : boolean;
var i, s : integer;
begin
  s := 1;
  result:=false;
  for i:=low(allAllergens) to high(allAllergens) do
  begin
    if allAllergens[i]=item then
      if ((s and score) <>0 ) then 
        exit(true);
    s := s shl 1; 
  end;
end;


end.
